## Welcome to **Gold Standard**!
This notebook provides a walkthrough for importing, tidying, and visualizing data from the [**2008 Summer Olympic Games**](https://edjnet.github.io/OlympicsGoNUTS/2008/) in Beijing. Install the necessary dependencies and download adapted data by following the instructions on [**Github**](https://github.com/annamccartan3/MCCARTAN-Data-Science-Portfolio/tree/main/TidyData-Project). Then return to explore the top performers of 2008!

### Load Libraries & Data

In [114]:
# import necessary libraries
import pandas as pd

# import untidy dataset
olympic_df = pd.read_csv('data/olympics08.csv')

### Data Exploration
- Examine the features included in this file
- View summary statistics for the dataset
- Check for missing values

In [115]:
# view the first 5 rows of the df
olympic_df.head()

medalist_name male_archery female_archery male_athletics  \
0    Aaron Armstrong          NaN            NaN           gold   
1      Aaron Peirsol          NaN            NaN            NaN   
2   Abdullo Tangriev          NaN            NaN            NaN   
3  Abeer Abdelrahman          NaN            NaN            NaN   
4            Abhinav          NaN            NaN            NaN   

  female_athletics male_badminton female_badminton male_baseball  \
0              NaN            NaN              NaN           NaN   
1              NaN            NaN              NaN           NaN   
2              NaN            NaN              NaN           NaN   
3              NaN            NaN              NaN           NaN   
4              NaN            NaN              NaN           NaN   

  male_basketball female_basketball  ... female_beach volleyball  \
0             NaN               NaN  ...                     NaN   
1             NaN               NaN  ...                     NaN   
2             NaN               NaN  ...                     NaN   
3             NaN               NaN  ...                     NaN   
4             NaN               NaN  ...                     NaN   

  male_volleyball female_volleyball male_water polo female_water polo  \
0             NaN               NaN             NaN               NaN   
1             NaN               NaN             NaN               NaN   
2             NaN               NaN             NaN               NaN   
3             NaN               NaN             NaN               NaN   
4             NaN               NaN             NaN               NaN   

  male_weightlifting female_weightlifting male_freestyle wrestling  \
0                NaN                  NaN                      NaN   
1                NaN                  NaN                      NaN   
2                NaN                  NaN                      NaN   
3                NaN               bronze                      NaN   
4                NaN                  NaN                      NaN   

  female_freestyle wrestling male_greco-roman wrestling  
0                        NaN                        NaN  
1                        NaN                        NaN  
2                        NaN                        NaN  
3                        NaN                        NaN  
4                        NaN                        NaN  

[5 rows x 71 columns]

In [132]:
print(f"Summary Statistics:")

# check total number of athletes/events
print(f"Total athletes: {len(olympic_df)}")
print(f"Total events: {len(olympic_df.columns)}")

# view summary statistics
olympic_df.describe()

Summary Statistics:
Total athletes: 1875
Total events: 71


medalist_name male_archery female_archery male_athletics  \
count              1875           11              9             85   
unique             1875            3              3              3   
top     Aaron Armstrong       bronze         bronze         silver   
freq                  1            4              4             31   

       female_athletics male_badminton female_badminton male_baseball  \
count                81             12               10            72   
unique                3              3                3             3   
top              silver         silver           silver        silver   
freq                 28              4                4            24   

       male_basketball female_basketball  ... female_beach volleyball  \
count               35                36  ...                       6   
unique               3                 3  ...                       3   
top             bronze            bronze  ...                    gold   
freq                12                12  ...                       2   

       male_volleyball female_volleyball male_water polo female_water polo  \
count               36                36              39                39   
unique               3                 3               3                 3   
top             bronze              gold          silver              gold   
freq                12                12              13                13   

       male_weightlifting female_weightlifting male_freestyle wrestling  \
count                  24                   21                       28   
unique                  3                    3                        3   
top                  gold               bronze                   bronze   
freq                    8                    7                       14   

       female_freestyle wrestling male_greco-roman wrestling  
count                          16                         27  
unique                          3                          3  
top                        bronze                     bronze  
freq                            8                         13  

[4 rows x 71 columns]

In [ ]:
# check for missing values
print("Missing Values:")
olympic_df.isnull().sum()

Missing Values:


medalist_name                    0
male_archery                  1864
female_archery                1866
male_athletics                1790
female_athletics              1794
                              ... 
male_weightlifting            1851
female_weightlifting          1854
male_freestyle wrestling      1847
female_freestyle wrestling    1859
male_greco-roman wrestling    1848
Length: 71, dtype: int64

#### Initial Observations:
We se that in the raw dataset, each row is an olympian and each column is an event. A total of 1875 athletes and 71 events are recorded. For each athlete, all values are missing except for events where the athlete earned a medal and the medal type is listed. This data is categorical: for each event, the athletes earned either gold ('gold'), silver ('silver'), bronze ('bronze'), or no medal (NaN).

### Data Tidying

#### In tidy data:
- Each variable forms a column
- Each observation forms a row
- Each type of observational unit forms a table

**There are several factors that make the current dataset untidy**:
- Column headers are values, not variable names
    - Right now, there is a separate column for every single sport featured in the dataset (the data is "wide"). We want to instead create a single column for the variable "event" (we need to create a "long" dataset by melting the dataframe).
- Multiple variables are stored in one column
    - Each event has both a "male" and "female" iteration, but this information is stored in the same column, meaning that we are unable to aggregate by just one variable. We can fix this by splitting the column into two separate variables: event and gender.

In [143]:
# use pd.melt() to convert "wide" data to "long" data
events_list = list(olympic_df.columns)[1:]
olympic_df_melted = pd.melt(olympic_df,
                            id_vars="medalist_name",
                            value_vars=events_list,
                            var_name="Event",
                            value_name="Medal")

# use str.split() to separate the 'event' column into men's vs. women's events
olympic_df_melted[['Gender', 'Event']] = olympic_df_melted['Event'].str.split('_', expand=True)

# examine the tidied dataset
olympic_df_clean = olympic_df_melted.copy()
olympic_df_clean

medalist_name                  Event Medal Gender
0         Aaron Armstrong                archery   NaN   male
1           Aaron Peirsol                archery   NaN   male
2        Abdullo Tangriev                archery   NaN   male
3       Abeer Abdelrahman                archery   NaN   male
4                 Abhinav                archery   NaN   male
...                   ...                    ...   ...    ...
131245        Éver Banega  greco-roman wrestling   NaN   male
131246  Ólafur Stefánsson  greco-roman wrestling   NaN   male
131247      Óscar Brayson  greco-roman wrestling   NaN   male
131248   Łukasz Pawłowski  greco-roman wrestling   NaN   male
131249        Živko Gocić  greco-roman wrestling   NaN   male

[131250 rows x 4 columns]


### Visualization

### Pivot-Table
We can use the pivot_table() function to aggregate the tidied data and compare total medals earned across various events and genders. 

In [140]:
# use pivot_table() to aggregate by gender, event, and medal type
olympic_df_clean.pivot_table(index="Event",
                              columns=["Gender", "Medal"],
                              values="medalist_name",
                              aggfunc="count")

Gender                female                male             
Medal                 bronze  gold silver bronze  gold silver
Event                                                        
archery                  4.0   2.0    3.0    4.0   3.0    4.0
artistic gymnastics      7.0   7.0    6.0   12.0   7.0   10.0
association football    18.0  18.0   18.0   18.0  19.0   18.0
athletics               26.0  27.0   28.0   29.0  25.0   31.0
badminton                3.0   3.0    4.0    4.0   4.0    4.0
baseball                 NaN   NaN    NaN   24.0  24.0   24.0
basketball              12.0  12.0   12.0   12.0  11.0   12.0
beach volleyball         2.0   2.0    2.0    2.0   2.0    2.0
bmx                      1.0   1.0    1.0    1.0   1.0    1.0
boxing                   NaN   NaN    NaN   22.0  11.0   11.0
canoeing and kayaking    8.0   7.0    6.0   19.0  17.0   17.0
diving                   6.0   2.0    5.0    5.0   5.0    6.0
equestrian sport         9.0   4.0    7.0    5.0   7.0    6.0
fencing                  5.0  11.0   10.0   10.0   7.0   11.0
field hockey            16.0  16.0   16.0   17.0  17.0   16.0
freestyle wrestling      8.0   4.0    4.0   14.0   7.0    7.0
greco-roman wrestling    NaN   NaN    NaN   13.0   7.0    7.0
handball                14.0  14.0   14.0   14.0  15.0   14.0
judo                    14.0   7.0    7.0   14.0   7.0    7.0
modern pentathlon        NaN   1.0    1.0    1.0   1.0    1.0
mountain biking          1.0   1.0    1.0    1.0   1.0    1.0
rhythmic gymnastics      7.0   7.0    7.0    NaN   NaN    NaN
road bicycle racing      2.0   2.0    2.0    2.0   1.0    2.0
rowing                  18.0  20.0   20.0   27.0  28.0   28.0
sailing                  7.0   7.0    7.0   11.0  11.0   11.0
shooting sport           6.0   6.0    5.0    9.0   8.0    9.0
softball                15.0  15.0   15.0    NaN   NaN    NaN
swimming                22.0  21.0   20.0   21.0  18.0   23.0
synchronized swimming   11.0   9.0    9.0    NaN   NaN    NaN
table tennis             4.0   1.0    4.0    4.0   1.0    4.0
taekwondo                8.0   4.0    4.0    8.0   4.0    4.0
tennis                   3.0   3.0    3.0    3.0   3.0    3.0
track cycling            3.0   3.0    3.0   13.0   8.0   14.0
trampoline gymnastics    1.0   1.0    1.0    1.0   1.0    1.0
triathlon                1.0   1.0    1.0    1.0   1.0    1.0
volleyball              12.0  12.0   12.0   12.0  12.0   12.0
water polo              13.0  13.0   13.0   13.0  13.0   13.0
weightlifting            7.0   7.0    7.0    8.0   8.0    8.0

### Export Cleaned Data

In [ ]:
# export tidied dataset
# olympic_df_clean.to_csv('data/olympics08_cleaned.csv')